src/utils.ts

```diff
@@ -1108,3 +1073,38 @@ export function hasAtomicOperators(doc: Document | Document[]): boolean {
   const keys = Object.keys(doc);
   return keys.length > 0 && keys[0][0] === '$';
 }
+
+/**
+ * Merge inherited properties from parent into options, prioritizing values from options,
+ * then values from parent.
+ * @internal
+ */
+export function resolveOptions<T extends CommandOperationOptions>(
+  parent: OperationParent | undefined,
+  options?: T
+): T {
+  const result: T = Object.assign({}, options);
+  const session = options?.session;
+
+  // Users cannot pass a readConcern/writeConcern to operations in a transaction
+  if (!session?.inTransaction()) {
+    const readConcern = ReadConcern.fromOptions(options) ?? parent?.readConcern;
+    if (readConcern) {
+      result.readConcern = readConcern;
+    }
+
+    const writeConcern = WriteConcern.fromOptions(options) ?? parent?.writeConcern;
+    if (writeConcern) {
+      result.writeConcern = writeConcern;
+    }
+  }
+
+  const readPreference = ReadPreference.fromOptions(options) ?? parent?.readPreference;
+  if (readPreference) {
+    result.readPreference = readPreference;
+  }
+
+  const bsonOptions = resolveBSONOptions(result, parent);
+
+  return { ...result, ...bsonOptions };
```

Comment:
I think we can reduce allocations a bit by resolving the BSON objects early, otherwise you're assigning all of them using `Object.assign` above just to spread them here again. In this case `resolveBSONOptions` is more like a "santize data + provide defaults", so what if we did:
```
const result: T = Object.assign({ ...resolveBSONOptions(options, parent) }, options, );
```

Second observation: I'm sure I used `Object.assign` to get around issues with typing, but it may be worth investigating whether it's possible to gain stricter typing by spreading here for all options:
```
const result: T = { ...options, ...resolveBSONOptions(options, parent) };
```